### PR TITLE
[CUDA] Bugfix: Improve NVCC cuda_std flags handling when it differs from cpp_std

### LINKED ISSUE #8523: Combining default_options : ['cpp_std=c++2a', 'cuda_std=c++17'] breaks nvcc
I'm trying to compile a project where the host code (*.cpp) uses c++20, but the device code (*.cu) uses c++17, since that is the latest version currently supported by nvcc.

**Describe the bug**
In `meson.build`, I tried this:
```
project('C++ std', 'cpp', 'cuda', version : '1.0.0', default_options : ['cpp_std=c++2a', 'cuda_std=c++17'])
```
but `nvcc` fails because it was called with `--std=c++17 -std=c++2a`:

```
meson compile -v
ninja: Entering directory `.'
[1/2] nvcc -Xcompiler=-Wall -Xcompiler=-Winvalid-pch -Xcompiler=-Wnon-virtual-dtor --std=c++17 -std=c++2a -g -G -g -I.. -I. -Iprog.p -Iprog.p -o prog.p/main.cu.o -c ../main.cu
FAILED: prog.p/main.cu.o 
nvcc -Xcompiler=-Wall -Xcompiler=-Winvalid-pch -Xcompiler=-Wnon-virtual-dtor --std=c++17 -std=c++2a -g -G -g -I.. -I. -Iprog.p -Iprog.p -o prog.p/main.cu.o -c ../main.cu
nvcc fatal   : Value 'c++2a' is not defined for option 'std'
ninja: build stopped: subcommand failed.
```

**To Reproduce**
Start with cuda example [6 std](https://github.com/mesonbuild/meson/tree/master/test%20cases/cuda/6%20std), but use this:
```project('C++ std', 'cpp', 'cuda', version : '1.0.0', default_options : ['cpp_std=c++2a', 'cuda_std=c++17'])```

**Expected behavior**
I expected that `gcc` would be invoked with c++2a and `nvcc` would be invoked with c++17.

If I manually remove the spurious `-std=c++2a` then `nvcc` succeeds.

**system parameters**
This is a native build on 5.4.101-1-MANJARO with Meson 0.57.1.

### PR DESCRIPTION
**Pull request by obilaniu**
Closes #8523 , plus minor cleanups.

@dcbaker, can this also be marked for backport to 0.57.2?

### CODE REVIEW

**Review by dcbaker — APPROVED** [review]

File: mesonbuild/compilers/cuda.py [hunk 1]
```diff
@@ -616,8 +616,19 @@ def get_options(self) -> 'KeyedOptionDictType':
...
+
+        # We must strip the -std option from the host compiler option set, as NVCC has
+        # its own -std flag that may not agree with the host compiler's.
         overrides = {name: opt.value for name, opt in options.items()}
-        return OptionOverrideProxy(overrides, self.host_compiler.get_options())
+        overrides.pop(OptionKey('std', machine=self.for_machine,
+                                       lang=self.host_compiler.language), None)
+        host_options = self.host_compiler.get_options().copy()
+        if 'std' in host_options:
+            del host_options['std']
```
Comment (obilaniu) [inline mesonbuild/compilers/cuda.py:630]: @dcbaker I need some help here with Mypy type annotations. How do I express the idea that I want to delete the `std` option both from the host compiler and from the overrides, that special flag being handled already beginning at line 638?

Comment (obilaniu) [reply mesonbuild/compilers/cuda.py:630]: @dcbaker How should I handle the above?

Comment (jpakkane) [reply mesonbuild/compilers/cuda.py:630]: Unless someone can come up with a better solution soon, disabling mypy for that line is fine.

**Comment by jpakkane** [comment]
Merged as #8596 so closing.

**Pull request closed without merging**